File: fsm.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>

using namespace std;

// a "table" with the key being (current state, next word) and
//the value as the next state to go to
map<pair<int, string>, int> table;

void initialization(){
    //create a table with all the valid states and next states
    table.insert(make_pair(make_pair(0, "the"), 1));
    table.insert(make_pair(make_pair(1, "lazy"), 1));
    table.insert(make_pair(make_pair(1, "smelly"), 1));
    table.insert(make_pair(make_pair(1, "dog"), 2));
    table.insert(make_pair(make_pair(1, "cat"), 2));
    table.insert(make_pair(make_pair(2, "ate"), 3));
    table.insert(make_pair(make_pair(2, "ran"), 3));
    table.insert(make_pair(make_pair(3, "slowly"), 4));
    table.insert(make_pair(make_pair(3, "noisily"), 4));
    table.insert(make_pair(make_pair(3, "EOI"), 5));
    table.insert(make_pair(make_pair(4, "EOI"), 5));
}

bool valid(vector<string> sentence){
    //initial state is 0
    int state = 0;
    for(int i = 0; i < sentence.size(); i++){
        //create a pair called key that contains the state and the next word
        pair<int, string> key = make_pair(state, sentence[i]);
        //if key is not found, the sentence is not valid, otherwise continue 
        //iterating over the sentence, getting the next state
        if(table.find(key) == table.end())
            return false;
        else
            state = table[key];
    }
    //if the final state is 5, my end state, and the word in the sentence is EOI, its valid
    if(state == 5 && sentence[sentence.size() - 1] == "EOI")
        return true;
    return false;
}

void tests(){
    //a variable to store each test sentence in
    vector<vector<string> > testSentences {
        {"the", "dog", "ate", "EOI"},
        {"the", "cat", "ate", "EOI"},
        {"the", "smelly", "dog", "ate", "EOI"},
        {"the", "lazy", "cat", "ran", "EOI"},
        {"the", "smelly", "cat", "ate", "noisily", "EOI"},
        {"the", "lazy", "dog", "ate", "slowly", "EOI"},
        {"the", "lazy", "smelly", "smelly", "smelly", "dog", "ate", "EOI"},
        {"the", "lazy", "dog", "barked", "noisily", "EOI"},
        {"the", "dog", "ate"},
        {"the", "EOI"}
    };
    //loop over every test sentence and test if its valid
    for(int i = 0; i < testSentences.size(); i++){
        if(valid(testSentences[i]))
            cout << "Sentence " << i << " is valid" << endl;
        else
            cout << "Sentence " << i << " is invalid" << endl;
    }
}

int main() {
    initialization();
    tests();
    return 0;
}

File: fsm.py
# Name: John Park
# ID:   446 88 167

table = {}
#   state  word    next possible states
table[(0, 'the'    )] = [1, 2]
table[(1, 'lazy'   )] = [1, 2]
table[(1, 'smelly' )] = [1, 2]
table[(2, 'dog'    )] = [3]
table[(2, 'cat'    )] = [3]
table[(3, 'ate'    )] = [4]
table[(3, 'ran'    )] = [4]
table[(4, 'slowly' )] = [5]
table[(4, 'noisily')] = [5]
table[(4, 'EOI'    )] = ['END']
table[(5, 'EOI'    )] = ['END']

def isValid(sentence):
    state = 0
    words = sentence.split(' ')
    for index, word in enumerate(words):
        try:
            # reached END with valid input
            if table[(state, word)][0]=='END' and (state==4 or state==5): 
                return True
            
            # reached end of sentence without reaching END
            elif word==words[-1]: 
                return False
            
            # move to next state
            else:
                next = table[(state, word)]
                if len(next)==1: state = next[0]
                else:
                    if words[index + 1]=='dog' or words[index + 1]=='cat': 
                        state = next[1]
                    else: 
                        state = next[0]

        # invalid state and word input
        except:
            return False

def test():
    assert(isValid('the lazy smelly dog ate slowly EOI' ))
    assert(isValid('the smelly lazy smelly cat ran EOI' ))
    assert(isValid('the smelly dog ate noisily EOI'     ))
    assert(isValid('the lazy cat ran noisily EOI'       ))
    assert(isValid('the lazy smelly cat ran EOI'        ))
    assert(isValid('the lazy dog ate slowly EOI'        ))
    assert(isValid('the cat ran EOI'                    ))
    assert(isValid('the dog ate EOI'                    ))
    assert(isValid('the lazy lazy smelly dog ran')==False)
    assert(isValid('the smelly cat ran noisily'  )==False)
    assert(isValid('the cat ate ate EOI'         )==False)
    assert(isValid('the ate slowly EOI'          )==False)
    assert(isValid('dog ate EOI'                 )==False)
    assert(isValid('cat ran'                     )==False)
    assert(isValid('the EOI'                     )==False)
    assert(isValid('the'                         )==False)
    assert(isValid('EOI'                         )==False)
    assert(isValid(''                            )==False)
    print('All tests passed')

def main():
    test()

if __name__ == '__main__':
    main()